### PR TITLE
Fixes for pipeline multi sample issue 2159 and 2180

### PR DESCRIPTION
**Pull request by jzulauf-lunarg**
Fixes for https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/issues/2180 and https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/issues/2159

Includes a common code refactor of the CreateGraphicsPipeline setup code for layer_validation_tests.cpp, used by the new cases, but presented here for review/feedback/refinement before refactoring the 20+ other copies in extant test cases to use it.

### CODE REVIEW

File: tests/layer_validation_tests.cpp [hunk 1]
```diff
@@ -173,13 +176,18 @@ struct NearestGreaterThan {
 };
 template <typename T, typename Comp>
 T NearestNonEqual(const T value, const Comp &comp) {
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: The previous version had some numerical corner cases uncovered when expanding the usage.  This is simpler and while it has an iterative fallback (for further missed corners), I couldn't find a test case that uses it.  There are some ULP differences between this and the prior implementation that don't matter to layer validation testing.

File: tests/layer_validation_tests.cpp [hunk 2]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
         return pool_ != VK_NULL_HANDLE && layout_ != VK_NULL_HANDLE && set_ != VK_NULL_HANDLE;
     }
 };
+
+struct CreatePipelineHelper {
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: This refactors out common setup code for testing CreateGraphicsPipelines.  For typical usage see OneshotTest below

File: tests/layer_validation_tests.cpp [hunk 3]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        }
+        if (implicit_destroy && (pipeline_ != VK_NULL_HANDLE)) {
+            vkDestroyPipeline(layer_test_->device(), pipeline_, nullptr);
+            pipeline_ = VK_NULL_HANDLE;
+        }
+        err = vkCreateGraphicsPipelines(layer_test_->device(), pipeline_cache_, 1, &gp_ci_, NULL, &pipeline_);
+        return err;
+    }
+    template <typename Test, typename OverrideFunc, typename Error>
+    static void OneshotTest(Test *test, OverrideFunc &info_override, const VkFlags flags, Error error, bool positive_test = false) {
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: The usage here is similar to the rest of the framework, init, twiddle, init some more, test.  Oneshot works for all the cases where simple twiddling with the create info structs (and the things they point to) in info_override is sufficient.

File: tests/layer_validation_tests.cpp [hunk 4]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        helper.CreateGraphicsPipeline();
+
+        if (positive_test) {
+            test->Monitor()->VerifyNotFound();
+        } else {
+            test->Monitor()->VerifyFound();
+        }
+    };
+};
+namespace chain_util {
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: More common code removal for a pattern in the pNext test added below

File: tests/layer_validation_tests.cpp [hunk 5]
```diff
@@ -9044,6 +9347,84 @@ TEST_F(VkLayerTest, InvalidPipelineCreateState) {
...
+    VkPhysicalDeviceFeatures device_features = {};
+    ASSERT_NO_FATAL_FAILURE(GetPhysicalDeviceFeatures(&device_features));
+    device_features.sampleRateShading = VK_FALSE;
+
+    ASSERT_NO_FATAL_FAILURE(InitState(&device_features));
+    ASSERT_NO_FATAL_FAILURE(InitRenderTarget());
+
+    // Cause the error by enabling sample shading...
+    auto set_shading_enable = [](CreatePipelineHelper &helper) { helper.pipe_ms_state_ci_.sampleShadingEnable = VK_TRUE; };
+    CreatePipelineHelper::OneshotTest(this, set_shading_enable, VK_DEBUG_REPORT_ERROR_BIT_EXT, VALIDATION_ERROR_10000620);
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: Typical OneshotTest usage.  The info_override is a simple one-liner.

File: tests/layer_validation_tests.cpp [hunk 6]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+    ~CreatePipelineHelper() {
+        VkDevice device = layer_test_->device();
+        vkDestroyPipelineCache(device, pipeline_cache_, nullptr);
+        vkDestroyPipeline(device, pipeline_, nullptr);
+        vkDestroyPipelineLayout(device, pipeline_layout_, nullptr);
+        vkDestroyDescriptorSetLayout(device, ds_layout_, nullptr);
+        vkDestroyDescriptorPool(device, ds_pool_, nullptr);
+    }
+
+    void InitDescriptorPoolInfo() {
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Reuse OneOffDescriptorSet?

File: tests/layer_validation_tests.cpp [hunk 7]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+    VkPipelineDynamicStateCreateInfo dyn_state_ci_ = {};
+    std::vector<VkPipelineShaderStageCreateInfo> shader_stages_;
+    VkPipelineRasterizationStateCreateInfo rs_state_ci_ = {};
+    VkPipelineColorBlendAttachmentState cb_attachments_ = {};
+    VkPipelineColorBlendStateCreateInfo cb_ci_ = {};
+    VkGraphicsPipelineCreateInfo gp_ci_ = {};
+    VkPipelineCacheCreateInfo pc_ci_ = {};
+    VkPipeline pipeline_ = VK_NULL_HANDLE;
+    VkPipelineCache pipeline_cache_ = VK_NULL_HANDLE;
+    std::unique_ptr<VkShaderObj> vs_;
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Why separate VS/FS members? The tests you want to replace use a mix of stages.

File: tests/layer_validation_tests.cpp [hunk 8]
```diff
@@ -946,6 +956,266 @@ struct OneOffDescriptorSet {
...
+
+struct CreatePipelineHelper {
+   public:
+    std::vector<VkDescriptorPoolSize> ds_pool_sizes_;
+    VkDescriptorPoolCreateInfo ds_pool_ci_ = {};
+    VkDescriptorPool ds_pool_ = VK_NULL_HANDLE;
+    std::vector<VkDescriptorSetLayoutBinding> dsl_bindings_;
+    VkDescriptorSetLayoutCreateInfo ds_layout_ci_ = {};
+    VkDescriptorSetLayout ds_layout_ = VK_NULL_HANDLE;
+    VkDescriptorSet descriptorSet_ = VK_NULL_HANDLE;
```
Comment (tobine) [inline tests/layer_validation_tests.cpp]: descriptor_set_

File: tests/layer_validation_tests.cpp (hunk 6 above)
Comment (jzulauf-lunarg) [reply tests/layer_validation_tests.cpp]: Looking further into the test cases, VkPipelineObj should be something I'm using instead as well.

File: tests/layer_validation_tests.cpp (hunk 7 above)
Comment (jzulauf-lunarg) [reply tests/layer_validation_tests.cpp]: Most are VS/FS only, but looking at adapting/replacing the helper with use of the VkPipelineObj (which I didn't notice before), and that supports more stages AFAICT

File: tests/layer_validation_tests.cpp [hunk 9]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        //    VkPipelineVertexInputStateCreateInfo
+        //    VkPipelineInputAssemblyStateCreateInfo
+        //    VkPipelineViewportStateCreateInfo
+        //    VkPipelineRasterizationStateCreateInfo
+        //    VkPipelineMultisampleStateCreateInfo
+        //    VkPipelineColorBlendStateCreateInfo
+        gp_ci_.sType = VK_STRUCTURE_TYPE_GRAPHICS_PIPELINE_CREATE_INFO;
+        gp_ci_.pNext = nullptr;
+        gp_ci_.flags = VK_PIPELINE_CREATE_DISABLE_OPTIMIZATION_BIT;
+        gp_ci_.pVertexInputState = &vi_ci_;
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: `stageCount` and `pStages` not set here. Should at least note that.

File: tests/layer_validation_tests.cpp [hunk 10]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        gp_ci_.pInputAssemblyState = &ia_ci_;
+        gp_ci_.pTessellationState = nullptr;
+        gp_ci_.pViewportState = &vp_state_ci_;
+        gp_ci_.pRasterizationState = &rs_state_ci_;
+        gp_ci_.pMultisampleState = &pipe_ms_state_ci_;
+        gp_ci_.pDepthStencilState = nullptr;
+        gp_ci_.pColorBlendState = &cb_ci_;
+        // If pDynamicState is pointing to dyn_state_ci_, it will late bind to dyn_states_ in LateBindCreatePipeline
+        gp_ci_.pDynamicState = &dyn_state_ci_;
+        gp_ci_.renderPass = layer_test_->renderPass();
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: rest seems mostly explicitly initialized here. should say `subpass = 0`?

File: tests/layer_validation_tests.cpp [hunk 11]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        InitDescriptorPoolInfo();
+        InitDescriptorSetInfo();
+        InitInputAndVertexInfo();
+        InitMultisampleInfo();
+        InitPipelineLayoutInfo();
+        InitViewportInfo();
+        InitDynamicStateInfo();
+        InitShaderInfo();
+        InitRasterizationInfo();
+        InitBlendStateInfo();
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: These should preferably be in order they appear in `VkGraphicsPipelineCreateInfo`.

File: tests/layer_validation_tests.cpp [hunk 12]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        InitGraphicsPipelineInfo();
+        InitPipelineCacheInfo();
+    }
+
+    void InitState() {
+        VkResult err;
+        ds_pool_ci_.poolSizeCount = ds_pool_sizes_.size();  // Late bind from vector, to allow modification
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: Should be protected from 0 size.

File: tests/layer_validation_tests.cpp [hunk 13]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        ds_pool_ci_.pPoolSizes = ds_pool_sizes_.data();
+        err = vkCreateDescriptorPool(layer_test_->device(), &ds_pool_ci_, NULL, &ds_pool_);
+        ASSERT_VK_SUCCESS(err);
+
+        ds_layout_ci_.bindingCount = dsl_bindings_.size();  // Late bind from vector, to allow modifications
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: Should be protected from 0 size.

File: tests/layer_validation_tests.cpp [hunk 14]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        InitBlendStateInfo();
+        InitGraphicsPipelineInfo();
+        InitPipelineCacheInfo();
+    }
+
+    void InitState() {
+        VkResult err;
+        ds_pool_ci_.poolSizeCount = ds_pool_sizes_.size();  // Late bind from vector, to allow modification
+        ds_pool_ci_.pPoolSizes = ds_pool_sizes_.data();
+        err = vkCreateDescriptorPool(layer_test_->device(), &ds_pool_ci_, NULL, &ds_pool_);
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: does not really allow to set `ds_pool_`. But `ds_pool_` is public! Should only create if it is `VK_NULL_HANDLE` ?

File: tests/layer_validation_tests.cpp [hunk 15]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+
+        // Handle value in info struct, need to late bind here
+        alloc_info_.descriptorPool = ds_pool_;
+        err = vkAllocateDescriptorSets(layer_test_->device(), &alloc_info_, &descriptorSet_);
+        ASSERT_VK_SUCCESS(err);
+
+        err = vkCreatePipelineLayout(layer_test_->device(), &pipeline_layout_ci_, NULL, &pipeline_layout_);
+        ASSERT_VK_SUCCESS(err);
+
+        err = vkCreatePipelineCache(layer_test_->device(), &pc_ci_, NULL, &pipeline_cache_);
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: I wonder if it helps anything...

File: tests/layer_validation_tests.cpp [hunk 16]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
         return pool_ != VK_NULL_HANDLE && layout_ != VK_NULL_HANDLE && set_ != VK_NULL_HANDLE;
     }
 };
+
+struct CreatePipelineHelper {
+   public:
+    std::vector<VkDescriptorPoolSize> ds_pool_sizes_;
+    VkDescriptorPoolCreateInfo ds_pool_ci_ = {};
+    VkDescriptorPool ds_pool_ = VK_NULL_HANDLE;
+    std::vector<VkDescriptorSetLayoutBinding> dsl_bindings_;
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: could just be private static array?
the user can use his own vector, or modify the default array through the `ds_layout_ci_.pBindings`.

File: tests/layer_validation_tests.cpp [hunk 17]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        ASSERT_VK_SUCCESS(err);
+
+        err = vkCreatePipelineLayout(layer_test_->device(), &pipeline_layout_ci_, NULL, &pipeline_layout_);
+        ASSERT_VK_SUCCESS(err);
+
+        err = vkCreatePipelineCache(layer_test_->device(), &pc_ci_, NULL, &pipeline_cache_);
+        ASSERT_VK_SUCCESS(err);
+    }
+
+    void LateBindPipelineInfo() {
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: Should be private? The `CreateGraphicsPipeline` seems to call it implicitly anyway, so probably no need to call it explicitly

File: tests/layer_validation_tests.cpp [hunk 18]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+            dyn_state_ci_.pDynamicStates = dyn_states_.data();
+        }
+    }
+
+    VkResult CreateGraphicsPipeline(bool implicit_destroy = true, bool do_late_bind = true) {
+        VkResult err;
+        if (do_late_bind) {
+            LateBindPipelineInfo();
+        }
+        if (implicit_destroy && (pipeline_ != VK_NULL_HANDLE)) {
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: `vkDestroy*` should be able to destroy `VK_NULL_HANDLE`. Though there used to be some driver bugs, so probably better.

File: tests/layer_validation_tests.cpp [hunk 19]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        gp_ci_.stageCount = shader_stages_.size();
+        gp_ci_.pStages = shader_stages_.data();
+        if (gp_ci_.pDynamicState == &dyn_state_ci_) {
+            // Late bind the dyn_states_ array
+            dyn_state_ci_.dynamicStateCount = dyn_states_.size();
+            dyn_state_ci_.pDynamicStates = dyn_states_.data();
+        }
+    }
+
+    VkResult CreateGraphicsPipeline(bool implicit_destroy = true, bool do_late_bind = true) {
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: `implicit_destroy` is bit confusing name. I did not really expect it to do that. Maybe `destroy_previous_pipeline`. Even so it would be weird the old pipeline to just get leaked.
Maybe just create the pipelines into a stack?

File: tests/layer_validation_tests.cpp [hunk 20]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        return err;
+    }
+    template <typename Test, typename OverrideFunc, typename Error>
+    static void OneshotTest(Test *test, OverrideFunc &info_override, const VkFlags flags, Error error, bool positive_test = false) {
+        CreatePipelineHelper helper(test);
+        helper.InitInfo();
+        info_override(helper);
+        helper.InitState();
+
+        test->Monitor()->SetDesiredFailureMsg(flags, error);
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: Does this work?
The positive tests usually set something like `m_errorMonitor->ExpectSuccess();`.

File: tests/layer_validation_tests.cpp [hunk 21]
```diff
@@ -946,6 +958,297 @@ struct OneOffDescriptorSet {
...
+        shader_stages_[0] = vs_->GetStageCreateInfo();
+        shader_stages_[1] = fs_->GetStageCreateInfo();
+    }
+
+    void InitRasterizationInfo() {
+        rs_state_ci_.sType = VK_STRUCTURE_TYPE_PIPELINE_RASTERIZATION_STATE_CREATE_INFO;
+        rs_state_ci_.pNext = nullptr;
+        rs_state_ci_.flags = 0;
+        rs_state_ci_.depthClampEnable = VK_FALSE;
+        rs_state_ci_.rasterizerDiscardEnable = VK_FALSE;
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: This may be very commonly used setting. Might be worth to make some shortcut for setting `VK_TRUE`.

File: tests/layer_validation_tests.cpp (hunk 1 above)
Comment (jzulauf-lunarg) [reply tests/layer_validation_tests.cpp]: Converting to std::nextafter

Comment (krOoze) [reply tests/layer_validation_tests.cpp]: @jzulauf-lunarg From experience, it wont pass the Android CI here. You migt have to use the C version of the function.

File: layers/vk_validation_error_database.txt [hunk 22]
```diff
@@ -1143,12 +1143,12 @@ VALIDATION_ERROR_0fe1c40d~^~Y~^~Unknown~^~vkCreatePipelineLayout~^~VUID-VkPipeli
...
+VALIDATION_ERROR_1001c40d~^~Y~^~InvalidPipelineSamplePNext~^~vkCmdSetViewport~^~VUID-VkPipelineMultisampleStateCreateInfo-pNext-pNext~^~core~^~The spec valid usage text states 'Each pNext member of any structure (including this one) in the pNext chain must be either NULL or a pointer to a valid instance of VkPipelineCoverageModulationStateCreateInfoNV, VkPipelineCoverageToColorStateCreateInfoNV, or VkPipelineSampleLocationsStateCreateInfoEXT' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-VkPipelineMultisampleStateCreateInfo-pNext-pNext)~^~implicit, TBD in parameter validation layer.
```
[the remark this answers is not in the document]
Comment (jzulauf-lunarg) [reply layers/vk_validation_error_database.txt]: There are improvements needed that will best be done in enhancements to code generation

**Comment by jzulauf-lunarg** [comment]
Can I get a thumbs up or a set of requested changes?

**Comment by mark-lunarg** [comment]
LGTM.
